core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -110,7 +110,12 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
       throw new IllegalArgumentException("Only literal resources are supported. Got: " + resource.patternType)
     }
 
-    val principal = session.principal
+    // ensure we compare identical classes
+    val principal = if (classOf[KafkaPrincipal] != session.principal.getClass)
+      new KafkaPrincipal(session.principal.getPrincipalType, session.principal.getName)
+    else
+      session.principal
```

Comment:
We could add a new local val `sessionPrincipal` to hold `session.principal` since we are using it four times here.